### PR TITLE
[CausalLM/layers] implement SwiGLU backward pass for LoRA training

### PR DESCRIPTION
**Pull request by sumon-98**
## Dependency of the PR
None — standalone layer change, no dependency on other PRs in this series.

## Commits to be reviewed in this PR

<details><summary>[CausalLM/layers] implement SwiGLU backward pass for LoRA training</summary><br />

Implements `forwarding()` (previously a no-op) and `calcDerivative()` in
`SwiGLULayer` to enable gradient flow during LoRA fine-tuning.

- `finalize()`: request `sigmoid_gate` scratch tensor (ITERATION_LIFESPAN) to cache sigmoid values for reuse in backprop
- `forwarding()`: now implemented — delegates to `incremental_forwarding(0, height, training)` so training and inference share a single code path
- `incremental_forwarding()`: minor rename `_from` → `step` for clarity
- `calcDerivative()`: full Swish′ gradient:
  - `d_up   = Swish(gate) × dy`
  - `d_gate = Swish′(gate) × up × dy`, where `Swish′(g) = σ(g) + g·σ(g)·(1−σ(g))`

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Niket Agarwal <niket.a@samsung.com>
Signed-off-by: Sumon Nath <sumon.nath@samsung.com>

</details>

### Summary

- Add `calcDerivative()` to `SwiGLULayer` — full Swish′ gradient with cached `sigmoid_gate` scratch tensor
- Implement `forwarding()` which was previously a no-op, unifying training and inference paths
- Enables SwiGLU to participate in backpropagation for LoRA fine-tuning of Qwen3

Signed-off-by: Niket Agarwal <niket.a@samsung.com>
Signed-off-by: Sumon Nath <sumon.nath@samsung.com>

### CODE REVIEW

**Comment by baek2sm** [comment]
please rebase it to main :)

**Review by Jungwon-Lee — COMMENTED** [review]
**Test coverage**

No unit test is included for the new `calcDerivative()` — only "Build/Run test passed" is checked in the self-eval. Swish-derivative math like this can regress silently (build/run still pass, gradients still wrong). Could we add a test comparing the analytic gradient against a numerical (finite-difference) gradient for `gate`/`up`?

**Review by Jungwon-Lee — COMMENTED** [review]
**Nit: `tensor_idx` array is overkill for a single scratch tensor**

This layer only ever requests one scratch tensor (`sigmoid_gate`), so the `std::array<unsigned int, 1> tensor_idx` + `enum SwiGLUParams { sigmoid_gate = 0 }` combo is more machinery than needed. The array+enum pattern in this codebase (e.g. `layer_normalization_layer.h`) is for layers with *multiple* scratch tensors; for a single one, the existing convention (e.g. `pooling2d_layer.h`'s `pool_helper_idx`) is just a plain `unsigned int` member. Suggested simplification:

```cpp
// swiglu.h
-#include <array>
+#include <limits>
 ...
 private:
-  std::array<unsigned int, 1> tensor_idx;
+  unsigned int sigmoid_gate_idx = std::numeric_limits<unsigned int>::max();
 
 public:
   WIN_EXPORT SwiGLULayer() :
-    Layer(), swiglu_props(nntrainer::props::SkipPrefill()) {
-    tensor_idx.fill(std::numeric_limits<unsigned>::max());
-  }
+    Layer(), swiglu_props(nntrainer::props::SkipPrefill()) {}
 ...
 private:
   std::tuple<nntrainer::props::SkipPrefill> swiglu_props;
   bool skip_prefill = false;
-  enum SwiGLUParams { sigmoid_gate = 0 };
 };
```

```cpp
// swiglu.cpp
-  tensor_idx[SwiGLUParams::sigmoid_gate] =
+  sigmoid_gate_idx =
     context.requestTensor(context.getInputDimensions()[0], "sigmoid_gate",
                           nntrainer::Initializer::NONE, false,
                           nntrainer::TensorLifespan::ITERATION_LIFESPAN);
 ...
-  nntrainer::Tensor &sig_gate =
-    context.getTensor(tensor_idx[SwiGLUParams::sigmoid_gate]);
+  nntrainer::Tensor &sig_gate = context.getTensor(sigmoid_gate_idx);
```

Not blocking, just simpler and drops the unused `<array>` include.

**Review by Jungwon-Lee — CHANGES_REQUESTED** [review]
Requesting changes because the new `sigmoid_gate` scratch tensor is allocated only in `finalize()` and is not resized by `updateTensorsByInputDimensions()`. Dynamic sequence-length changes can therefore produce a shape mismatch or an out-of-bounds write during backward. Please resize the cache with the inputs/output and add a finite-difference gradient test that also exercises dimension changes.
